File: LeetCode_Algorithm/206(Reverse Linked List)/206.py
from typing import List, Optional

from numpy import true_divide

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        # sol1 O(n), O(1) 36~73ms 91~10%
        root = self.makeTree(head,0)     

        if head is None:
            return head

        
        curr = head.next
        prev = head
        prev.next = None
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        return prev


    def makeTree(self, head, idx):
         rootNode = ListNode(head[idx])
         if (idx+1 < len(head)): rootNode.next = self.makeTree(head, idx+1)
         return rootNode



print(Solution().reverseList([1,2,3,4,5]))

File: LeetCode_Algorithm/206(Reverse Linked List)/206.js
 function ListNode(val, next) {
     this.val = (val===undefined ? 0 : val)
     this.next = (next===undefined ? null : next)
 }

var reverseList = function(head) {
  // sol1 O(n), O(1) 56~76ms 99~70%
  if (head == null) {return null}
  if (head.next == undefined){return head}
  var root = makeTree(head, 0)
  
  var curr = root.next
  var prev = root
  prev.next = null
  while (curr != null) {
    var temp = curr.next
    curr.next = prev
    prev = curr
    curr = temp
  }
  return prev
};
var makeTree = function(head, idx){
  var rootNode = new ListNode(head[idx])
  if(idx+1<head.length){rootNode.next = makeTree(head, idx+1)}
  return rootNode
}
console.log(reverseList([]))